src/GraphQL.ApiTests/netstandard20+netstandard21/GraphQL.approved.txt

```diff
@@ -2668,7 +2669,7 @@ namespace GraphQL.Validation
         public System.Collections.Generic.IEnumerable<GraphQLParser.AST.ASTNode> Nodes { get; }
         public string Number { get; set; }
     }
-    public class ValidationOptions
+    public readonly struct ValidationOptions
```

Comment:
Why changing to struct?  Why readonly struct?  Why setters if readonly struct? MS guidelines state to avoid a struct unless it contains < 16 bytes, and this struct is 144 bytes.

https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/choosing-between-class-and-struct

Reply:
In the code I see you're using init-only properties.  Are you sure this works properly with older compilers that do not recognize the init-only attributes?  Since the struct is readonly and has no constructors, it seems like it could be a major problem.

Reply:
> Why changing to struct?

To save allocation for such transitive object as options. 

> Why readonly struct? 

To be able to use `in` modifier to pass struct by reference without guard copy. Fixed in latest commit. I just forgot to use `in`.

> Why setters if readonly struct?

Because `init` != `set`.

Reply:
> Are you sure this works properly with older compilers that do not recognize the init-only attributes?

I'm 100% sure that this feature does not work for older compilers. Of course.

>  it seems like it could be a major problem.

Whose problem?

Reply:
We target .NET Standard 2.0 so people can use whatever type of environment they wish. Mono or VS, webassembly, .NET Framework, .NET Core, C#, F# whatever. We shouldn’t use features that require the newest software if we are targeting old frameworks.

I actually think I remember that init-only is fine with older compilers — but perhaps not when the struct is marked as read only. There is a document on it from MS but I’m not sure where it is at the moment.

Reply:
I have multiple active projects today that run on old or outdated frameworks. One project I was working on yesterday was a website in Classic ASP. We don’t support that, but we should make a conscience decision if we are breaking support for older frameworks or compilers, and then list it in the migration notes. 

I think it would be a shame to drop support for older tool sets or frameworks for the sake of one allocation per request. Especially when doing something as simple as adding a constructor with all the parameters can satisfy the needs of such a toolset.

Reply:
Old runtime != old compiler. `init` property is a compiler feature. Old runtime **can** use classes compiled with `init` properties, it just a `set` property. Regarding `readonly struct` - i'm not 100% sure what happens. Well, `readonly` is a compiler feature as well so I think old compilers can just set any value of "readonly struct". Try and check if you work with outdated frameworks.